File: COURSE 1 (Algorithmic Toolbox)/week3_greedy_algorithms/0.Week 3 Assignment/src/com/company/CoveringSegments.java
package com.company;

import java.util.*;

public class CoveringSegments {

    static final Comparator<Segment> COMPARISON_SEGMENTS= new Comparator<>() {
        @Override
        public int compare(Segment o1, Segment o2) {
            return Long.compare(o1.end, o2.end);
        }
    };

    //o(nlogn)
    private static long[] optimalPoints(Segment[] segments) {

        Arrays.sort(segments,COMPARISON_SEGMENTS);

        List<Long> result=new LinkedList<>();
        int i=0;
        result.add(segments[0].end);
        long p=segments[0].end;

        for (i = 1; i < segments.length; i++) {

            if(segments[i].start>p){
                p=segments[i].end;
                result.add(p);
            }
        }

        long[] points=new long[result.size()];

        for(i=0;i<result.size();i++){
            points[i]=result.get(i);
        }

        return points;
    }

    private static class Segment {
        long start, end;

        Segment(long start, long end) {
            this.start = start;
            this.end = end;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        Segment[] segments = new Segment[n];
        for (int i = 0; i < n; i++) {
            long start, end;
            start = scanner.nextLong();
            end = scanner.nextLong();
            segments[i] = new Segment(start, end);
        }
        long[] points = optimalPoints(segments);
        System.out.println(points.length);
        for (long point : points) {
            System.out.print(point + " ");
        }
    }
}

File: COURSE 1 (Algorithmic Toolbox)/week3_greedy_algorithms/0.Week 3 Assignment/submitted files c++/covering_segments.cpp
#include <algorithm>
#include <iostream>
#include <climits>
#include <vector>

using namespace std;
using std::vector;

struct Segment {
  int start, end;
};

bool comparison(const Segment &a,const Segment &b){
  return a.end<b.end;
}

vector<int> optimal_points(vector<Segment> &segments) {
    sort(segments.begin(),segments.end(),comparison);
    vector<int> points;
    vector<int> result;
    result.push_back(segments[0].end);
    int p=segments[0].end;

    for (int i = 1; i < segments.size(); i++) {
       if(segments[i].start>p){
        p=segments[i].end;
        result.push_back(p);
     }
  }

  for(int i = 0; i < result.size(); i++){
    points.push_back(result[i]);
  }
  return points;
}

int main() {
  int n;
  std::cin >> n;
  vector<Segment> segments(n);
  for (int i = 0; i < segments.size(); i++) {
    std::cin >> segments[i].start >> segments[i].end;
  }
  vector<int> points = optimal_points(segments);
  std::cout << points.size() << "\n";
  for (int i = 0; i < points.size(); i++) {
    std::cout << points[i] << " ";
  }
}
